### PR TITLE
fix: highlight appropriate navItem based on active sidebar item

### PR DESCRIPTION
**Pull request by tapanchudasama**
## Motivation

This PR fixes https://github.com/facebook/docusaurus/issues/6268

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?
Yes

## Test Plan
Verified the changes by checking the color of the navitem.

### Current Behaviour (on docusaurus.io)
image

`Docs` navitem should be highlighted.

### After these changes:
image
`Docs` navitem is highlighted as expected.

### CODE REVIEW

**Review by Josh-Cena — APPROVED** [review]
The current behavior looks fine to me, but not sure if there could be edge cases...

**Review by Josh-Cena — CHANGES_REQUESTED** [review]
Actually, tests are failing because this has led to behavior changes. Every doc must match its versioned counterparts with the exact same path.

**Comment by tapanchudasama** [comment]
> Actually, tests are failing because this has led to behavior changes. 

Yes, I came to know when I ran `yarn test` locally but before modifying tests, I want to know if this is this the way to fix the bug I am addressing? Asking because I am first-time contributor to this project.

**Comment by Josh-Cena** [comment]
Unfortunately this makes the version dropdown lose its context. Before, navigation through the versions dropdown would take you to the same page in a different version; now, you are always taken to the doc homepage. The routes need to be matched exactly to determine the version context.

I suspect it's actually the doc navbar item's active condition that needs to be changed, but can't be sure either... 🤔

**Comment by Josh-Cena** [comment]
I just looked at it. I think the actual problem is that category generated indexes are not added to the global data and hence `activeDoc.sidebar` doesn't exist, so the actual fix would be to modify `plugin-content-docs`, edit `globalData`, and make the category index docs registered within the global version. Sorry for the misinformation!

If you have any questions, just ask. I can help you push it through.

**Comment by tapanchudasama** [comment]
> so the actual fix would be to modify plugin-content-docs, edit globalData, and make the category index docs registered within the global version. 

Oh, okay. I see. I'll see what I can do. And let you know, if there are any things that I need to know.

Thanks btw!

**Comment by slorber** [comment]
> I just looked at it. I think the actual problem is that category generated indexes are not added to the global data and hence activeDoc.sidebar doesn't exist, so the actual fix would be to modify plugin-content-docs, edit globalData, and make the category index docs registered within the global version. 

that seems correct 👍

**Comment by tapanchudasama** [comment]
Okay, So I did as you said. But and its also solved the issue. Now how to fix this TS issue? @Josh-Cena ? Right now, I have just ignored it.

**Comment by Josh-Cena** [comment]
Thanks @tapanchudasama I'll fix this myself. TS relies on control-flow analysis so you will need to first filter out the case of `doc` being a doc item, and then it will allow you to access the properties of a subtype.

File: packages/docusaurus-plugin-content-docs/src/globalData.ts [hunk 1]
```diff
@@ -24,6 +28,16 @@ export function toGlobalDataDoc(doc: DocMetadata): GlobalDoc {
   };
 }
 
+export function toGlobalDataGeneratedIndex(
+  doc: CategoryGeneratedIndexMetadata,
+): GlobalDoc {
+  return {
+    id: doc.slug,
```
Comment (Josh-Cena) [inline packages/docusaurus-plugin-content-docs/src/globalData.ts:35]: This is fine. We never hook into a doc's ID in our built-in hooks, and user can only call hooks on generated index when they swizzle `DocCategoryGeneratedIndex`. At this point, it's already known that the doc you are receiving is a generated index, so using the `id` doesn't make sense.

**Comment by Josh-Cena** [comment]
Going to merge as the behavior seems all correct now. Thanks @tapanchudasama

File: packages/docusaurus-plugin-content-docs/src/globalData.ts [hunk 2]
```diff
@@ -56,7 +70,9 @@ export function toGlobalDataVersion(version: LoadedVersion): GlobalVersion {
     isLast: version.isLast,
     path: version.versionPath,
     mainDocId: version.mainDocId,
-    docs: version.docs.map(toGlobalDataDoc),
+    docs: version.docs
+      .map(toGlobalDataDoc)
+      .concat(version.categoryGeneratedIndices.map(toGlobalDataGeneratedIndex)),
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/globalData.ts:75]: I don't like this part much 😅 generated index are not docs, why not create an extra separate attribute instead of trying to put these in the same array?

The items do not need to have a weird id=slug in such a case

Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/globalData.ts:75]: On the contrary, those are different things, and the differences should rather be handled in client-side hooks.

If I use `useDocs("xyz")`, and there is a doc with id=xyz + a generated index with slug=xyz, which one should be returned? 

These edge cases can happen

Comment (Josh-Cena) [reply packages/docusaurus-plugin-content-docs/src/globalData.ts:75]: So for one thing, all slugs are prefixed with `/`, so there would never be clashes between doc IDs and slugs

Comment (Josh-Cena) [reply packages/docusaurus-plugin-content-docs/src/globalData.ts:75]: I understand that this is not very useful and probably confusing, but it improves uniformity while won't introduce any bugs. Maybe we want to formalize "generated index IDs"?

Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/globalData.ts:75]: > I understand that this is not very useful and probably confusing, but it improves uniformity while won't introduce any bugs. 

We can keep it for now as it's likely to work fine in most cases, but I'd like to change that later

> Maybe we want to formalize "generated index IDs"?

I don't know if this is needed 🤷‍♂️ a slug/permalink should already be unique, the id is just a convenience for the user to reference in sidebar/navbar etc
